cmd/util/ledger/migrations/account_based_migration.go

```diff
@@ -1,189 +1,316 @@
+	}
+	close(jobs)
+
+	// make sure to exit all workers before returning from this function
+	// so that the migrator can be closed properly
+	wg.Wait()
+
+	log.Info().
+		Array("top_longest_migrations", durations.Array()).
+		Msgf("Top longest migrations")
+
+	if err != nil {
+		return nil, fmt.Errorf("fail to migrate payload: %w", err)
 	}
 
 	return migrated, nil
 }
 
 type jobMigrateAccountGroup struct {
-	Account  string
+	Address  common.Address
 	Payloads []ledger.Payload
 }
 
 type migrationResult struct {
+	migrationDuration
+
 	Migrated []ledger.Payload
 	Err      error
 }
 
-// MigrateGroupConcurrently migrate the payloads in the given payloadsByAccount map which
-// using the migrator
-// It's similar to MigrateGroupSequentially, except it will migrate different groups concurrently
-func MigrateGroupConcurrently(
-	migrator AccountMigrator,
-	payloadsByAccount map[string][]ledger.Payload,
-	nWorker int,
-) (
-	[]ledger.Payload, error) {
+// payloadToAccount takes a payload and return:
+// - (address, true, nil) if the payload is for an account, the account address is returned
+// - ("", false, nil) if the payload is not for an account
+// - ("", false, err) if running into any exception
+func payloadToAccount(p ledger.Payload) (common.Address, bool, error) {
+	k, err := p.Key()
+	if err != nil {
+		return common.Address{}, false, fmt.Errorf("could not find key for payload: %w", err)
+	}
 
-	jobs := make(chan jobMigrateAccountGroup, len(payloadsByAccount))
-	go func() {
-		for account, payloads := range payloadsByAccount {
-			jobs <- jobMigrateAccountGroup{
-				Account:  account,
-				Payloads: payloads,
-			}
-		}
-		close(jobs)
-	}()
+	id, err := util.KeyToRegisterID(k)
+	if err != nil {
+		return common.Address{}, false, fmt.Errorf("error converting key to register ID")
+	}
 
-	resultCh := make(chan *migrationResult)
-	for i := 0; i < int(nWorker); i++ {
-		go func() {
-			for job := range jobs {
-				accountMigrated, err := migrator.MigratePayloads(job.Account, job.Payloads)
-				resultCh <- &migrationResult{
-					Migrated: accountMigrated,
-					Err:      err,
-				}
-			}
-		}()
+	if len([]byte(id.Owner)) != flow.AddressLength {
+		return common.Address{}, false, nil
 	}
 
-	// read job results
-	logAccount := util.LogProgress("processing account group", len(payloadsByAccount), &log.Logger)
+	address, err := common.BytesToAddress([]byte(id.Owner))
+	if err != nil {
+		return common.Address{}, false, fmt.Errorf("invalid account address: %w", err)
+	}
 
-	migrated := make([]ledger.Payload, 0)
+	return address, true, nil
+}
 
-	for i := 0; i < len(payloadsByAccount); i++ {
-		result := <-resultCh
-		if result.Err != nil {
-			return nil, fmt.Errorf("fail to migrate payload: %w", result.Err)
-		}
+// payloadGroup groups Payloads by account.
+// For global Payloads, it's stored under NonAccountPayloads field
+type payloadGroup struct {
+	NonAccountPayloads []ledger.Payload
+	Accounts           map[common.Address][]ledger.Payload
```

Comment:
Maybe replace `Payload` with `*Payload` to reduce each payload element size from 48 bytes -> 8 bytes.

```suggestion
	NonAccountPayloads []*ledger.Payload
	Accounts           map[common.Address][]*ledger.Payload
```